src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -18,192 +18,181 @@ internal sealed class TaskCenterSolutionAnalysisProgressReporter
-                _resettableDelay.Task.SafeContinueWith(_ =>
+                // Kick off task to update the UI after a delay to pick up any new events.
+                _intervalTask = Task.Delay(s_minimumInterval).ContinueWith(_ =>
                 {
-                    _resettableDelay = null;
-                    ProgressUpdated();
+                    ReportProgress();
                 }, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
             }
         }
 
-        private void StartedOrStopped(bool started)
+        private void ReportProgress()
         {
-            if (started)
+            lock (_lock)
             {
-                ResetProgressStatus();
-
-                // if there is any pending one. make sure it is finished.
-                _currentTask?.TrySetResult(default);
-
-                var taskHandler = _taskCenterService.PreRegister(_options, data: default);
-
-                _currentTask = new TaskCompletionSource<VoidResult>();
-                taskHandler.RegisterTask(_currentTask.Task);
+                var data = _lastProgressData;
+                _intervalTask = null;
 
-                // report initial progress
-                ChangeProgress(taskHandler, message: null);
+                _updateUITask = _updateUITask.ContinueWith(_ => UpdateUI(data),
+                    CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+            }
+        }
 
-                // set handler
-                _taskHandler = taskHandler;
+        private void UpdateUI(ProgressData progressData)
+        {
+            if (progressData.Status == ProgressStatus.Stopped)
+            {
+                StopTaskCenter();
+                return;
             }
-            else
+
+            // Update the pending item count if the progress data specifies a value.
+            if (progressData.PendingItemCount.HasValue)
             {
-                // clear progress message
-                ChangeProgress(_taskHandler, message: null);
+                _lastPendingItemCount = progressData.PendingItemCount.Value;
+            }
 
-                // stop progress
-                _currentTask?.TrySetResult(default);
-                _currentTask = null;
+            // Start the task center task if not already running.
+            if (_taskHandler == null)
+            {
+                StartTaskCenter();
+            }
 
-                _taskHandler = null;
+            var statusMessage = progressData.Status == ProgressStatus.Paused
+                ? ServicesVSResources.Paused_0_tasks_in_queue
+                : ServicesVSResources.Evaluating_0_tasks_in_queue;
 
-                ResetProgressStatus();
-            }
+            _taskHandler?.Progress.Report(new TaskProgressData
```

Comment:
```suggestion
            _taskHandler.Progress.Report(new TaskProgressData
```

Reply:
if it complains that _taskHandler may be null, an alternative method is to write things like so:

```c#
var taskHandler = _taskHandler ?? StartTaskCenter();
taskHandler.Progress.Report(...)
```

Reply:
note: i thnk this is strictly better because thre is no way that _taskHandler should be null here.  so having to write `?.` just makes things more confusing as it implies that null is expected here.

Reply:
it does warn, I think I'll just inline this method then, it should then be able to tell that _taskHandler won't be null.  

I guess nullable analysis thinks it is null b/c it's to expensive to keep exploring nested methods?

Reply:
C# in general, and NRT in specific, has no concept of inter-procedural analysis.